### PR TITLE
Adding support for extract

### PR DESCRIPTION
**Pull request by saurabhgupta-datametica**

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 1]
```diff
@@ -6836,6 +6836,62 @@ public void testCaseForLnFunction() {
       .ok(expectedBQ);
   }
 
+  @Test public void testExtractDecade() {
```
Comment (vaibhavjain-dm) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:6839]: Add test cases in SqlFunctionsTest as well.

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -495,6 +495,21 @@ private SqlNode makCastCallForTimeWithPrecision(SqlNode operandToCast, int preci
         super.unparseCall(writer, call, leftPrec, rightPrec);
       }
       break;
+    case EXTRACT:
+      if (call.operand(0).toString().equals("EPOCH")) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:499]: Extract to a method and refactor it using switch statements.

File: core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import org.apache.calcite.sql.type.SqlTypeUtil;
+
+import static org.apache.calcite.sql.fun.SqlStdOperatorTable.CAST;
+
+/**
+ * Support unparse logic for Extract function of Decade , Century , DOY , DOW
+ */
+public class ExtractFunctionFormatUtil {
+
+  public static final String DAY_OF_YEAR = "DOY";
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java:40]: Why public ?

File: core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      return handleExtractWithOperand(call.operand(1), DateTimeUnit.DAYOFYEAR);
+    case DAY_OF_WEEK:
+      return  handleExtractWithOperand(call.operand(1), DateTimeUnit.DAYOFWEEK);
+    case DECADE:
+      return handleExtractDeacade(call);
+    case CENTURY:
+      return handleExtractCentury(call);
+    case MILLENNIUM:
+      return handleExtractMillennium(call);
+
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java:59]: Please remove space.

File: core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    }
+    return null;
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java:61]: Do not return null.

File: core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            SqlParserPos.ZERO);
+    SqlNode[] substrOperand = new SqlNode[] { extractCall, divideLiteral};
+    SqlCall divideCall = new SqlBasicCall(SqlStdOperatorTable.DIVIDE, substrOperand,
+            SqlParserPos.ZERO);
+    SqlCall ceilCall = new SqlBasicCall(SqlStdOperatorTable.CEIL, new SqlNode[]{divideCall},
+            SqlParserPos.ZERO);
+    BasicSqlType sqlType = new BasicSqlType(RelDataTypeSystem.DEFAULT, SqlTypeName.INTEGER);
+    return CAST.createCall(SqlParserPos.ZERO, ceilCall, SqlTypeUtil.convertTypeToSpec(sqlType));
+  }
+  private SqlCall handleExtractMillennium(SqlCall call) {
```
Comment (vaibhavjain-dm) [inline core/src/main/java/org/apache/calcite/sql/dialect/ExtractFunctionFormatUtil.java:96]: handleExtractDeacade and handleExtractMillennium method are same refactor it.

**Comment by praveendatametica** [comment]
LGTM

**Review by soma-mondal — APPROVED** [review]

**Pull request merged**